src/compiler/compiler.rs

```diff
@@ -556,41 +560,44 @@ gcc
             //TODO: do something smarter here.
             if line == "gcc" {
                 debug!("Found GCC");
-                return future::ok(Some(CompilerKind::Gcc)).boxed()
+                return future::ok(Some(Box::new(CCompiler(GCC)) as Box<Compiler<T>>)).boxed()
             } else if line == "clang" {
                 debug!("Found clang");
-                return future::ok(Some(CompilerKind::Clang)).boxed()
+                return future::ok(Some(Box::new(CCompiler(Clang)) as Box<Compiler<T>>)).boxed()
             } else if line == "msvc" {
                 debug!("Found MSVC");
                 let prefix = msvc::detect_showincludes_prefix(&creator,
                                                               executable.as_ref(),
                                                               &pool);
                 return Box::new(prefix.map(|prefix| {
                     trace!("showIncludes prefix: '{}'", prefix);
-                    Some(CompilerKind::Msvc {
+                    Some(Box::new(CCompiler(MSVC {
                         includes_prefix: prefix,
-                    })
+                    })) as Box<Compiler<T>>)
                 }))
             }
         }
         future::ok(None).boxed()
     }))
 }
 
-/// If `executable` is a known compiler, return `Some(Compiler)` containing information about it.
+/// If `executable` is a known compiler, return a `CompilerInfo` containing information about it.
 pub fn get_compiler_info<T>(creator: &T, executable: &str, pool: &CpuPool)
-                            -> SFuture<Compiler>
+                            -> SFuture<CompilerInfo<T>>
     where T: CommandCreatorSync
 {
     let executable = executable.to_string();
     let pool = pool.clone();
-    Box::new(detect_compiler_kind(creator, &executable, &pool).and_then(move |kind| {
-        match kind {
-            Some(kind) => {
-                pool.spawn_fn(move || Compiler::new(&executable, kind))
-                    .chain_err(|| "failed to learn compiler metadata")
+    let detect = detect_compiler(creator, &executable, &pool);
+    Box::new(detect.and_then(move |compiler| {
+        match compiler {
+            Some(compiler) => {
+                Box::new(sha1_digest(executable.clone(), &pool)
```

Comment:
> This may also throw a wrench into sccache's own caching of the compiler as well unfortunately b/c rustup can have a different rustc depending on which directory it's invoked in (and that can change over time). I think sccache assumes the main compiler doesn't change much, right?

Also a good point, we'll have to do something smarter here. sccache caches compiler info by (path, mtime), so rustup will totally break this expectation. I'll have to think about that for a bit, I don't want to have an extra rustc invocation for every single compile if we can avoid it.